From 5ebf7eb8012b4aa23eb52b6bac99dfa7b9c1f1f9 Mon Sep 17 00:00:00 2001
From: Raymond Wong <infiwang@pm.me>
Date: Wed, 29 Mar 2023 15:40:53 +0800
Subject: [PATCH 072/156] riscv(ffi): fix BC_IS[EQ,NE]V vmeta cdata comparision
 trigger

Fix 5ab19d9 "riscv(ffi): FFI init"

This is a WIP, archive only.
---
 src/vm_riscv64.dasc | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/src/vm_riscv64.dasc b/src/vm_riscv64.dasc
index 7dd59af2..68d71439 100644
--- a/src/vm_riscv64.dasc
+++ b/src/vm_riscv64.dasc
@@ -2864,11 +2864,10 @@ static void build_ins(BuildCtx *ctx, BCOp op, int defop)
     |// Either or both types are not numbers.
     |.if FFI
     |  // Check if RA or RD is a cdata.
-    |  li TMP0, LJ_TCDATA
-    |  bne CARG3, TMP0, >1
-    |  bne CARG4, TMP0, >1
-    |  j ->vmeta_equal_cd
-    |1:
+    |  xori TMP0, CARG3, LJ_TCDATA
+    |  xori TMP1, CARG4, LJ_TCDATA
+    |  and TMP0, TMP0, TMP1
+    |  bxeqz TMP0, ->vmeta_equal_cd
     |.endif
     |  lui TMP3, (-(BCBIAS_J*4 >> 12)) & 0xfffff		// -BCBIAS_J*4
     |  decode_BC4b TMP2
-- 
2.42.0

